### PR TITLE
Adding spell checking for docs and spack root

### PR DESCRIPTION
**Pull request by vsoch**
Spack has a lot of spelling errors! And it's because we're human :laughing: But maybe we can help with some automation?

I found this tool typos (a rust library) and created a GitHub action for it this weekend. If we run it during CI, it can alert us to spelling errors! I am currently running for the spack docs and spack root, and we can add other spack subfolders when --exclude is supported (see issue https://github.com/crate-ci/typos/issues/269) 

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

### CODE REVIEW

**Comment by adamjstewart** [comment]
I'm amazed that this works so well without having thousands of false-positives.

File: .github/workflows/unit_tests.yaml [hunk 1]
```diff
@@ -53,6 +53,10 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
+    - name: Check for typos in docs
+      uses: crate-ci/typos@592b36d23c62cb378f6097a292bc902ee73f93ef # version 1.0.4
```
Comment (epage) [inline .github/workflows/unit_tests.yaml:57]: Is the `v1.0.4` tag not working or is this to protect against the tag being changed/removed?

Comment (vsoch) [reply .github/workflows/unit_tests.yaml:57]: It's best practice to use commits, even releases can technically be moving targets. See: https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#using-third-party-actions

>  Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps mitigate the risk of a bad actor adding a backdoor to the action's repository, as they would need to generate a SHA-1 collision for a valid Git object payload.

**Comment by vsoch** [comment]
@adamjstewart there were some false positives in the spack test folder trying to change hashes, and one in the util folder trying to change a variable name like fullpath to something else. If we can avoid checking code itself (and just docstrings) we could avoid this. I wasn't able to get `--exclude` working locally so I didn't add any subfolders for now. But docs and the spack root is a good start!

File: .github/workflows/unit_tests.yaml [hunk 2]
```diff
@@ -53,6 +53,10 @@ jobs:
...
+      with: 
+        files: ./lib/spack/docs/ ./lib/spack/spack/*.py
```
Comment (tgamblin) [inline .github/workflows/unit_tests.yaml:59]: I can't find it in the `crate-ci` docs -- does this just get the py files in `lib/spack/spack`, or is it recursive?  We also should include `lib/spack/llnl` and `var/spack/repos/builtin/packages/*/*.py`.

Comment (vsoch) [reply .github/workflows/unit_tests.yaml:59]: @tgamblin I added it to the maximum number of places that would have maximum benefit without making it harder for the developer. For example, if we add to lib/spack/spack, while it does find many spelling errors, it also targets variable names, e.g.,:

```python
error: `fullset` should be `fullest`
  --> lib/spack/spack/util/environment.py:1041:25
     |
1041 |     def set_intersection(fullset, *args):
```
This is intentionally 'fullset` (which is not a word but implies full set) and should be renamed to fullset. When I saw errors like this, I realized that adding it to spell check code would be a huge annoyance for the developer - we already have enough style checks that are hard to deal with (pending your recent changes of course).

But for documentation? That's the user facing thing, and it denotes how professional our project is. Arguably spelling mistakes in code won't be seen by most, just developers. But documentation will. So this explains my choice to add spell checking to the documentation but not to the code.

There could be some intermediate to just spell check (and show) errors in the code and then it's up to the developer to fix if he/she wants, but it wouldn't result in an error code. My suggestion thus is the following:

1. to get the action in with this current setup to handle the docs, this is priority
2. to have two different PRs for each of adding spell checking to packages and the code, each with possibly a different strategy depending on what is found (small number of cases we can add exceptions and let the check run, larger number of cases we can fix and have run but not make it fail a test).

Comment (epage) [reply .github/workflows/unit_tests.yaml:59]: > When I saw errors like this, I realized that adding it to spell check code would be a huge annoyance for the developer

I'm assuming the number of cases is high enough and growth is high enough to not be worth adding these to `typos.toml` file as "approved" spellings?

Comment (vsoch) [reply .github/workflows/unit_tests.yaml:59]: We could, but every time someone uses a variable that looks like a misspelled word it’s going to trigger an error.

**Pull request closed without merging**